integration-tests/src/upgrade/tester.rs

```diff
@@ -0,0 +1,352 @@
+
+        let l1_chain_admin = diamond_proxy.getAdmin().call().await?;
+        let l1_chain_admin =
+            interfaces::ChainAdmin::new(l1_chain_admin, tester.l1_provider.clone());
+
+        let bridgehub_owner = bridgehub.owner().call().await?;
+        let ctm_owner = ctm.owner().call().await?;
+        let diamond_proxy_admin = diamond_proxy.getAdmin().call().await?;
+        let l1_chain_admin_owner = l1_chain_admin.owner().call().await?;
+
+        // Bytecode supplier is a bit special: right now it's not discoverable
+        // The value is hardcoded, keep it aligned with `node/bin/src/config.rs`, it must correspond
+        // to the value stored in `zkos-l1-state.json`.
+        let bytecode_supplier_address = "0x883498218f553d748e48b43595a7d29a82939f01".parse()?;
+        anyhow::ensure!(
+            !tester
+                .l1_provider
+                .get_code_at(bytecode_supplier_address)
+                .await?
+                .is_empty(),
+            "Bytecode supplier contract is not deployed at expected address {bytecode_supplier_address:?}; if zkos-l1-state.json was updated, update the address in the test code"
+        );
+        let bytecode_supplier = interfaces::BytecodesSupplier::new(
+            bytecode_supplier_address,
+            tester.l1_provider.clone(),
+        );
+
+        Ok(Self {
+            tester,
+            bridgehub,
+            bridgehub_owner,
+            ctm,
+            ctm_owner,
+            diamond_proxy,
+            diamond_proxy_admin,
+            l1_chain_admin,
+            l1_chain_admin_owner,
+            bytecode_supplier,
+            protocol_version,
+        })
+    }
+
+    async fn enable_impersonation(&self) -> anyhow::Result<()> {
+        // Enable impersonation and fund all governance accounts
+        for addr in [
+            self.bridgehub_owner,
+            self.ctm_owner,
+            self.diamond_proxy_admin,
+            self.l1_chain_admin_owner,
+        ] {
+            self.tester
+                .l1_provider
+                .anvil_impersonate_account(addr)
+                .await?;
+            self.tester
+                .l1_provider
+                .send_transaction(
+                    TransactionRequest::default()
+                        .with_to(addr)
+                        .with_value(U256::from(10).pow(U256::from(18u64))), // 1 ETH
+                )
+                .await?
+                .expect_successful_receipt()
+                .await?;
+        }
+        Ok(())
+    }
+
+    async fn fund_wallets(&self) -> anyhow::Result<()> {
+        for wallet in [self.bridgehub_owner] {
+            self.tester
+                .l1_provider
+                .send_transaction(
+                    TransactionRequest::default()
+                        .with_to(wallet)
+                        .with_value(U256::from(10).pow(U256::from(18u64))), // 1 ETH
+                )
```

Comment:
Nit: there is a more direct `anvil_set_balance` available as well but not a big deal